nixos/modules/services/networking/pppd.nix

```
@@ -0,0 +1,137 @@
+          }));
+      };
+    };
+  };
+
+  config = let
+    # pppd likes to write directly into /var/run. This is rude on a
+    # modern system, so this wrapper transparently moves those files
+    # into /run/pppd.
+    pppd-wrapped = pkgs.stdenv.mkDerivation {
+      name = "pppd-wrapped";
+      phases = [ "installPhase" ];
+      buildInputs = [ pkgs.makeWrapper ];
+      installPhase = ''
+        mkdir -p $out/sbin
+        makeWrapper ${getBin config.services.pppd.package}/sbin/pppd $out/sbin/pppd \
+          --set LD_PRELOAD "${pkgs.libredirect}/lib/libredirect.so" \
+          --set NIX_REDIRECTS "/var/run=/run/pppd"
+      '';
+    };
+
+    enabledConfigs = filter (f: f.enable) (attrValues config.services.pppd.peers);
+
+    mkEtc = cfg: {
+      "ppp/peers/${cfg.name}".text = cfg.config;
+    };
+
+    mkSystemd = cfg: {
+      "pppd-${cfg.name}" = {
+        restartTriggers = [ config.environment.etc."ppp/peers/${cfg.name}".source ];
+        before = [ "network.target" ];
+        wants = [ "network.target" ];
+        after = [ "network-pre.target" ];
+        serviceConfig = {
+          ExecStart = "${getBin pppd-wrapped}/sbin/pppd call ${cfg.name} nodetach nolog";
+          Restart = "always";
+          RestartSec = 5;
+
+          AmbientCapabilities = "CAP_SYS_TTY_CONFIG CAP_NET_ADMIN CAP_NET_RAW CAP_SYS_ADMIN";
+          CapabilityBoundingSet = "CAP_SYS_TTY_CONFIG CAP_NET_ADMIN CAP_NET_RAW CAP_SYS_ADMIN";
+          LockPersonality = true;
+          MemoryDenyWriteExecute = true;
+          NoNewPrivileges = true;
+          PrivateMounts = true;
+          PrivateTmp = true;
+          ProtectControlGroups = true;
+          ProtectHome = true;
+          ProtectHostname = true;
+          ProtectKernelModules = true;
+          # pppd can be configured to tweak kernel settings.
+          ProtectKernelTunables = false;
+          ProtectSystem = "strict";
+          RemoveIPC = true;
+          RestrictAddressFamilies = "AF_PACKET AF_UNIX AF_PPPOX AF_ATMPVC AF_ATMSVC AF_INET AF_INET6 AF_IPX";
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          SecureBits = "no-setuid-fixup-locked noroot-locked";
+          SystemCallArchitectures = "native";
+
+          # All pppd instances on a system must share a runtime
+          # directory in order for PPP multilink to work correctly. So
+          # we give all instances the same /run/pppd directory to store
+          # things in.
+          #
+          # For the same reason, we can't set PrivateUsers=true, because
+          # all instances need to run as the same user to access the
+          # multilink database.
+          RuntimeDirectory = "pppd";
+          RuntimeDirectoryPreserve = true;
+        };
+        wantedBy = optional cfg.autostart "multi-user.target";
+      };
+    };
+
+    etcFiles = map mkEtc enabledConfigs;
+    systemdConfigs = map mkSystemd enabledConfigs;
+
+  in {
```

```suggestion
  in mkIf cfg.enable {
```

I think you should include a global enable option